File: repos/ex-litellm/test/ex_litellm/providers/groq_test.exs
defmodule ExLiteLLM.Providers.GroqTest do
  use ExUnit.Case, async: true

  alias ExLiteLLM.Providers.Adapter.Request
  alias ExLiteLLM.Providers.Groq

  defp req(params, model \\ "openai/gpt-oss-20b") do
    %Request{model: model, params: params, provider: :groq}
  end

  test "gpt-oss hides reasoning, aliases max_tokens, and floors completion budget" do
    body = Groq.transform_request(req(%{"max_tokens" => 64, "messages" => []}))
    assert body["model"] == "openai/gpt-oss-20b"
    assert body["include_reasoning"] == false
    assert body["max_completion_tokens"] == 2048
    refute Map.has_key?(body, "max_tokens")
  end

  test "coerces illegal reasoning_effort values instead of forwarding a Groq 400" do
    assert %{"reasoning_effort" => "high"} =
             Groq.transform_request(req(%{"reasoning_effort" => "max", "messages" => []}))

    assert %{"reasoning_effort" => "medium"} =
             Groq.transform_request(req(%{"reasoning_effort" => "auto", "messages" => []}))

    body = Groq.transform_request(req(%{"reasoning_effort" => "nope", "messages" => []}))
    refute Map.has_key?(body, "reasoning_effort")
  end

  test "strips Claude Code pattern fields Groq rejects as invalid regex" do
    body =
      Groq.transform_request(
        req(%{
          "messages" => [],
          "tools" => [
            %{
              "type" => "function",
              "function" => %{
                "name" => "Artifact",
                "parameters" => %{
                  "type" => "object",
                  "properties" => %{
                    "after" => %{
                      "type" => "string",
                      "pattern" => "^[A-Za-z0-9_=-]{1,4096}$"
                    }
                  }
                }
              }
            }
          ]
        })
      )

    after_schema = get_in(body, ["tools", Access.at(0), "function", "parameters", "properties", "after"])
    refute Map.has_key?(after_schema, "pattern")
    assert after_schema["type"] == "string"
  end

  test "non-gpt-oss models keep max_tokens and do not force include_reasoning" do
    body =
      Groq.transform_request(
        req(%{"max_tokens" => 64, "messages" => []}, "llama-3.1-8b-instant")
      )

    assert body["max_tokens"] == 64
    refute Map.has_key?(body, "include_reasoning")
    refute Map.has_key?(body, "max_completion_tokens")
  end

  test "chunk_parser reads delta.reasoning" do
    chunk =
      ExLiteLLM.Providers.OpenAICompatible.Shared.chunk_parser(%{
        "choices" => [
          %{
            "index" => 0,
            "delta" => %{"content" => "", "reasoning" => "hmm"},
            "finish_reason" => nil
          }
        ]
      })

    assert chunk.text == ""
    assert chunk.reasoning == "hmm"
  end
end

File: plugins/llm/run-claude/repos/go-litellm/internal/providers/groq_test.go
package providers

import "testing"

func TestShapeGroqGPTOss(t *testing.T) {
	out := ShapeGroq(map[string]any{"max_tokens": 64, "reasoning_effort": "max"}, "openai/gpt-oss-20b")
	if out["include_reasoning"] != false {
		t.Fatalf("include_reasoning: %v", out["include_reasoning"])
	}
	if out["max_completion_tokens"] != 2048 {
		t.Fatalf("max_completion_tokens: %v", out["max_completion_tokens"])
	}
	if _, ok := out["max_tokens"]; ok {
		t.Fatal("max_tokens should be aliased away")
	}
	if out["reasoning_effort"] != "high" {
		t.Fatalf("effort: %v", out["reasoning_effort"])
	}
	out = ShapeGroq(map[string]any{"reasoning_effort": "auto"}, "openai/gpt-oss-20b")
	if out["reasoning_effort"] != "medium" {
		t.Fatalf("auto: %v", out["reasoning_effort"])
	}
}

func TestShapeGroqDropsUnknownEffort(t *testing.T) {
	out := ShapeGroq(map[string]any{"reasoning_effort": "nope"}, "openai/gpt-oss-20b")
	if _, ok := out["reasoning_effort"]; ok {
		t.Fatalf("expected drop, got %v", out["reasoning_effort"])
	}
}

func TestShapeGroqStripsInvalidToolPatterns(t *testing.T) {
	body := map[string]any{
		"tools": []any{
			map[string]any{
				"type": "function",
				"function": map[string]any{
					"name": "Artifact",
					"parameters": map[string]any{
						"type": "object",
						"properties": map[string]any{
							"after": map[string]any{
								"type":    "string",
								"pattern": "^[A-Za-z0-9_=-]{1,4096}$",
							},
						},
					},
				},
			},
		},
	}
	out := ShapeGroq(body, "openai/gpt-oss-120b")
	tools := out["tools"].([]any)
	fn := tools[0].(map[string]any)["function"].(map[string]any)
	params := fn["parameters"].(map[string]any)
	props := params["properties"].(map[string]any)
	after := props["after"].(map[string]any)
	if _, ok := after["pattern"]; ok {
		t.Fatalf("pattern should be stripped: %+v", after)
	}
	if after["type"] != "string" {
		t.Fatalf("kept type: %+v", after)
	}
}

func TestShapeGroqLeavesNonGptOss(t *testing.T) {
	out := ShapeGroq(map[string]any{"max_tokens": 64}, "llama-3.1-8b-instant")
	if out["max_tokens"] != 64 {
		t.Fatalf("%v", out)
	}
	if _, ok := out["include_reasoning"]; ok {
		t.Fatal(out)
	}
}
